File: swift/Sources/EHBP/Derive.swift
import Foundation
import Crypto

/// Response key material for AES-256-GCM encryption/decryption (SPEC Section 4.4.1)
public struct ResponseKeyMaterial: Sendable {
    /// 32-byte AES-256 key
    public let key: SymmetricKey

    /// 12-byte base nonce, XORed with chunk index for each chunk
    public let nonceBase: Data

    public init(key: SymmetricKey, nonceBase: Data) {
        self.key = key
        self.nonceBase = nonceBase
    }
}

/// Derives response encryption keys from the HPKE exported secret (SPEC Section 4.4.1)
///
/// Implements OHTTP-style derivation:
/// ```
/// salt = concat(enc, response_nonce)
/// prk = HKDF-Extract(salt, secret)
/// aead_key = HKDF-Expand(prk, "key", 32)
/// aead_nonce = HKDF-Expand(prk, "nonce", 12)
/// ```
public func deriveResponseKeys(
    exportedSecret: Data,
    requestEnc: Data,
    responseNonce: Data
) throws -> ResponseKeyMaterial {
    guard exportedSecret.count == EHBPConstants.exportLength else {
        throw EHBPError.invalidInput("exported secret must be \(EHBPConstants.exportLength) bytes, got \(exportedSecret.count)")
    }
    guard requestEnc.count == EHBPConstants.requestEncLength else {
        throw EHBPError.invalidInput("request enc must be \(EHBPConstants.requestEncLength) bytes, got \(requestEnc.count)")
    }
    guard responseNonce.count == EHBPConstants.responseNonceLength else {
        throw EHBPError.invalidInput("response nonce must be \(EHBPConstants.responseNonceLength) bytes, got \(responseNonce.count)")
    }

    var salt = Data()
    salt.append(requestEnc)
    salt.append(responseNonce)

    let prk = HKDF<SHA256>.extract(inputKeyMaterial: SymmetricKey(data: exportedSecret), salt: salt)

    let keyData = HKDF<SHA256>.expand(
        pseudoRandomKey: prk,
        info: Data(EHBPConstants.responseKeyLabel.utf8),
        outputByteCount: EHBPConstants.aes256KeyLength
    )

    let nonceBase = HKDF<SHA256>.expand(
        pseudoRandomKey: prk,
        info: Data(EHBPConstants.responseNonceLabel.utf8),
        outputByteCount: EHBPConstants.aesGCMNonceLength
    )

    return ResponseKeyMaterial(
        key: keyData,
        nonceBase: nonceBase.withUnsafeBytes { Data($0) }
    )
}

/// Computes nonce for chunk: aead_nonce XOR seq (SPEC Section 4.4.2)
public func computeNonce(nonceBase: Data, seq: UInt64) -> Data {
    var nonce = Data(nonceBase)

    for i in 0..<8 {
        let shift = i * 8
        let byteIndex = EHBPConstants.aesGCMNonceLength - 1 - i
        nonce[byteIndex] ^= UInt8((seq >> shift) & 0xFF)
    }

    return nonce
}

/// Encrypts a chunk using AES-256-GCM. Returns ciphertext || tag.
public func encryptChunk(
    keyMaterial: ResponseKeyMaterial,
    seq: UInt64,
    plaintext: Data
) throws -> Data {
    let nonce = computeNonce(nonceBase: keyMaterial.nonceBase, seq: seq)

    let sealedBox = try AES.GCM.seal(
        plaintext,
        using: keyMaterial.key,
        nonce: try AES.GCM.Nonce(data: nonce)
    )

    return sealedBox.ciphertext + sealedBox.tag
}

/// Decrypts a chunk using AES-256-GCM. Expects ciphertext || tag format.
public func decryptChunk(
    keyMaterial: ResponseKeyMaterial,
    seq: UInt64,
    ciphertext: Data
) throws -> Data {
    let nonce = computeNonce(nonceBase: keyMaterial.nonceBase, seq: seq)

    guard ciphertext.count >= 16 else {
        throw EHBPError.decryptionFailed("ciphertext too short")
    }

    let tagStart = ciphertext.count - 16
    let encryptedData = ciphertext.prefix(tagStart)
    let tag = ciphertext.suffix(16)

    let sealedBox = try AES.GCM.SealedBox(
        nonce: try AES.GCM.Nonce(data: nonce),
        ciphertext: encryptedData,
        tag: tag
    )

    return try AES.GCM.open(sealedBox, using: keyMaterial.key)
}

/// Incrementally decrypts an EHBP length-prefixed response stream.
///
/// Feed encrypted network bytes with ``push(_:)`` and call ``finish()`` when
/// the source reaches EOF. Each returned chunk has been authenticated and can
/// be consumed before the source reaches EOF.
public struct ResponseDecryptor {
    private let keyMaterial: ResponseKeyMaterial
    private let maxChunkLength: Int
    private var buffer = [UInt8]()
    private var sequence: UInt64

    init(
        keyMaterial: ResponseKeyMaterial,
        maxChunkLength: Int = EHBPConstants.maxResponseChunkBytes
    ) {
        self.init(
            keyMaterial: keyMaterial,
            maxChunkLength: maxChunkLength,
            initialSequence: 0
        )
    }

    init(
        keyMaterial: ResponseKeyMaterial,
        maxChunkLength: Int,
        initialSequence: UInt64
    ) {
        self.keyMaterial = keyMaterial
        self.maxChunkLength = maxChunkLength
        self.sequence = initialSequence
    }

    /// Adds encrypted bytes and returns all newly authenticated plaintext chunks.
    public mutating func push(_ data: Data) throws -> [Data] {
        buffer.append(contentsOf: data)
        var plaintext = [Data]()

        while buffer.count >= 4 {
            let chunkLength = Int(buffer[0]) << 24 |
                              Int(buffer[1]) << 16 |
                              Int(buffer[2]) << 8 |
                              Int(buffer[3])

            if chunkLength == 0 {
                buffer.removeFirst(4)
                continue
            }
            if chunkLength > maxChunkLength {
                throw EHBPError.invalidResponse("response chunk exceeds maximum allowed size")
            }
            guard buffer.count >= 4 + chunkLength else {
                break
            }
            guard sequence < UInt64.max else {
                throw EHBPError.invalidResponse("response chunk sequence overflow")
            }

            let ciphertext = Data(buffer[4..<(4 + chunkLength)])
            buffer.removeFirst(4 + chunkLength)
            let opened = try decryptChunk(
                keyMaterial: keyMaterial,
                seq: sequence,
                ciphertext: ciphertext
            )
            sequence += 1
            plaintext.append(opened)
        }

        return plaintext
    }

    /// Validates that source EOF occurred on a frame boundary.
    public func finish() throws {
        guard buffer.isEmpty else {
            throw EHBPError.invalidResponse("truncated encrypted response chunk")
        }
    }
}

/// EHBP errors
public enum EHBPError: Error, LocalizedError {
    case invalidInput(String)
    case encryptionFailed(String)
    case decryptionFailed(String)
    case networkError(String)
    case invalidResponse(String)
    case missingHeader(String)

    public var errorDescription: String? {
        switch self {
        case .invalidInput(let msg): return "Invalid input: \(msg)"
        case .encryptionFailed(let msg): return "Encryption failed: \(msg)"
        case .decryptionFailed(let msg): return "Decryption failed: \(msg)"
        case .networkError(let msg): return "Network error: \(msg)"
        case .invalidResponse(let msg): return "Invalid response: \(msg)"
        case .missingHeader(let msg): return "Missing header: \(msg)"
        }
    }
}

File: python/src/ehbp/derive.py
"""Response key derivation and chunk framing (SPEC Sections 4.3 and 4.4).

Request bodies are sealed with HPKE directly (see :mod:`ehbp.identity`).
Response bodies use AES-256-GCM under keys derived from the request's HPKE
context via raw HKDF (RFC 5869), following OHTTP (RFC 9458):

    salt = enc || response_nonce
    prk  = HKDF-Extract(salt, exported_secret)
    key  = HKDF-Expand(prk, "key", 32)
    base = HKDF-Expand(prk, "nonce", 12)

The per-chunk nonce is ``base XOR seq`` with ``seq`` big-endian in the low 8
bytes. Raw HKDF and AES-GCM come from the audited ``cryptography`` package so
the derivation matches the Go/Rust/JS clients byte-for-byte.
"""

from __future__ import annotations

import struct
from dataclasses import dataclass

from cryptography.hazmat.primitives import hashes, hmac
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives.kdf.hkdf import HKDFExpand

from .errors import CryptoError, InvalidInputError, ProtocolError
from .protocol import (
    AES256_KEY_LENGTH,
    AES_GCM_NONCE_LENGTH,
    EXPORT_LENGTH,
    LENGTH_PREFIX_SIZE,
    MAX_CHUNK_LENGTH,
    MAX_SEQUENCE,
    REQUEST_ENC_LENGTH,
    RESPONSE_KEY_LABEL,
    RESPONSE_NONCE_LABEL,
    RESPONSE_NONCE_LENGTH,
)


@dataclass
class ResponseKeyMaterial:
    """Derived AES-256-GCM key and base nonce for one response."""

    key: bytes
    nonce_base: bytes

    def __repr__(self) -> str:
        return "ResponseKeyMaterial(key='[redacted]', nonce_base='[redacted]')"


def _hkdf_extract(salt: bytes, ikm: bytes) -> bytes:
    h = hmac.HMAC(salt, hashes.SHA256())
    h.update(ikm)
    return h.finalize()


def _hkdf_expand(prk: bytes, info: bytes, length: int) -> bytes:
    return HKDFExpand(algorithm=hashes.SHA256(), length=length, info=info).derive(prk)


def derive_response_keys(
    exported_secret: bytes, request_enc: bytes, response_nonce: bytes
) -> ResponseKeyMaterial:
    if len(exported_secret) != EXPORT_LENGTH:
        raise InvalidInputError(
            f"exported secret must be {EXPORT_LENGTH} bytes, got {len(exported_secret)}"
        )
    if len(request_enc) != REQUEST_ENC_LENGTH:
        raise InvalidInputError(
            f"request enc must be {REQUEST_ENC_LENGTH} bytes, got {len(request_enc)}"
        )
    if len(response_nonce) != RESPONSE_NONCE_LENGTH:
        raise InvalidInputError(
            f"response nonce must be {RESPONSE_NONCE_LENGTH} bytes, got {len(response_nonce)}"
        )

    salt = bytes(request_enc) + bytes(response_nonce)
    prk = _hkdf_extract(salt, bytes(exported_secret))
    key = _hkdf_expand(prk, RESPONSE_KEY_LABEL, AES256_KEY_LENGTH)
    nonce_base = _hkdf_expand(prk, RESPONSE_NONCE_LABEL, AES_GCM_NONCE_LENGTH)
    return ResponseKeyMaterial(key=key, nonce_base=nonce_base)


def compute_nonce(nonce_base: bytes, seq: int) -> bytes:
    if len(nonce_base) != AES_GCM_NONCE_LENGTH:
        raise InvalidInputError(f"nonce base must be {AES_GCM_NONCE_LENGTH} bytes")
    if seq < 0 or seq > MAX_SEQUENCE:
        raise ProtocolError("response chunk sequence out of range")
    nonce = bytearray(nonce_base)
    for i in range(8):
        nonce[AES_GCM_NONCE_LENGTH - 1 - i] ^= (seq >> (i * 8)) & 0xFF
    return bytes(nonce)


def encrypt_chunk(km: ResponseKeyMaterial, seq: int, plaintext: bytes) -> bytes:
    nonce = compute_nonce(km.nonce_base, seq)
    try:
        return AESGCM(km.key).encrypt(nonce, bytes(plaintext), b"")
    except Exception as err:  # noqa: BLE001 - normalize to a stable error shape
        raise CryptoError("failed to encrypt chunk") from err


def decrypt_chunk(km: ResponseKeyMaterial, seq: int, ciphertext: bytes) -> bytes:
    nonce = compute_nonce(km.nonce_base, seq)
    try:
        return AESGCM(km.key).decrypt(nonce, bytes(ciphertext), b"")
    except Exception as err:  # noqa: BLE001 - do not leak the failing crypto stage
        raise CryptoError("failed to decrypt chunk") from err


def frame_chunk(ciphertext: bytes) -> bytes:
    length = len(ciphertext)
    if length > MAX_CHUNK_LENGTH:
        raise InvalidInputError("ciphertext chunk is too large")
    return struct.pack(">I", length) + bytes(ciphertext)


def decrypt_framed_response(km: ResponseKeyMaterial, body: bytes) -> bytes:
    out = bytearray()
    offset = 0
    seq = 0
    total = len(body)

    while offset < total:
        if total - offset < LENGTH_PREFIX_SIZE:
            raise ProtocolError("truncated chunk length")
        chunk_len = struct.unpack_from(">I", body, offset)[0]
        offset += LENGTH_PREFIX_SIZE

        if chunk_len == 0:
            continue
        if total - offset < chunk_len:
            raise ProtocolError("truncated encrypted chunk")

        out += decrypt_chunk(km, seq, body[offset : offset + chunk_len])
        if seq >= MAX_SEQUENCE:
            raise ProtocolError("response chunk sequence overflow")
        seq += 1
        offset += chunk_len

    return bytes(out)


class FrameDecryptor:
    """Incrementally decrypts a length-prefixed AEAD stream.

    Feed network bytes with :meth:`push` (returns any fully decrypted plaintext
    chunks) and call :meth:`finish` at end-of-stream to detect a truncated
    trailing chunk. Zero-length frames are skipped without consuming a sequence
    number, matching the framing rules in SPEC Section 4.3.
    """

    def __init__(
        self, km: ResponseKeyMaterial, max_chunk_length: int = MAX_CHUNK_LENGTH
    ) -> None:
        self._km = km
        self._buffer = bytearray()
        self._seq = 0
        self._max_chunk_length = max_chunk_length

    def push(self, data: bytes) -> list[bytes]:
        self._buffer += data
        chunks: list[bytes] = []

        while True:
            if len(self._buffer) < LENGTH_PREFIX_SIZE:
                break
            chunk_len = struct.unpack_from(">I", self._buffer, 0)[0]
            if chunk_len == 0:
                del self._buffer[:LENGTH_PREFIX_SIZE]
                continue
            if chunk_len > self._max_chunk_length:
                raise ProtocolError("response chunk exceeds maximum allowed size")
            if len(self._buffer) < LENGTH_PREFIX_SIZE + chunk_len:
                break

            ciphertext = bytes(
                self._buffer[LENGTH_PREFIX_SIZE : LENGTH_PREFIX_SIZE + chunk_len]
            )
            del self._buffer[: LENGTH_PREFIX_SIZE + chunk_len]
            chunks.append(decrypt_chunk(self._km, self._seq, ciphertext))
            if self._seq >= MAX_SEQUENCE:
                raise ProtocolError("response chunk sequence overflow")
            self._seq += 1

        return chunks

    def finish(self) -> None:
        if self._buffer:
            raise ProtocolError("truncated encrypted response chunk")
